File: src/main/java/com/dec/decisland/particles/ParticleConfig.java
package com.dec.decisland.particles;

import net.minecraft.core.particles.ParticleType;
import net.minecraft.core.particles.SimpleParticleType;
import net.neoforged.neoforge.client.event.RegisterParticleProvidersEvent;

import java.util.function.Supplier;

public class ParticleConfig {
    public String name;
    public Supplier<ParticleType<?>> sup;
    public Supplier<SimpleParticleType> registeredType; // 添加字段保存注册后的粒子类型

    // 粒子工厂提供者配置
    public ParticleFactoryProvider factoryProvider;

    public ParticleConfig(String name, Supplier<ParticleType<?>> sup, ParticleFactoryProvider factoryProvider) {
        this.name = name;
        this.sup = sup;
        this.factoryProvider = factoryProvider;
    }

    public static class Builder {
        private String name;
        private Supplier<ParticleType<?>> sup = () -> new SimpleParticleType(true);
        private ParticleFactoryProvider factoryProvider;

        public Builder(String name) {
            this.name = name;
        }

        public Builder sup(Supplier<ParticleType<?>> sup) {
            this.sup = sup;
            return this;
        }

        public Builder factoryProvider(ParticleFactoryProvider factoryProvider) {
            this.factoryProvider = factoryProvider;
            return this;
        }

        public ParticleConfig build() {
            return new ParticleConfig(name, sup, factoryProvider);
        }
    }

    // 粒子工厂提供者接口
    public interface ParticleFactoryProvider {
        void register(RegisterParticleProvidersEvent event, SimpleParticleType particleType);
    }
}

File: src/main/java/com/dec/decisland/particles/ParticleConfig.kt
package com.dec.decisland.particles

import net.minecraft.core.particles.ParticleType
import net.minecraft.core.particles.SimpleParticleType
import net.neoforged.neoforge.client.event.RegisterParticleProvidersEvent
import java.util.function.Supplier

class ParticleConfig(
    @JvmField val name: String,
    @JvmField val sup: Supplier<ParticleType<*>>,
    @JvmField val factoryProvider: ParticleFactoryProvider?,
) {
    @JvmField
    var registeredType: Supplier<SimpleParticleType>? = null

    class Builder(@JvmField val name: String) {
        private var sup: Supplier<ParticleType<*>> = Supplier { SimpleParticleType(true) }
        private var factoryProvider: ParticleFactoryProvider? = null

        fun sup(sup: Supplier<ParticleType<*>>): Builder = apply {
            this.sup = sup
        }

        fun factoryProvider(factoryProvider: ParticleFactoryProvider): Builder = apply {
            this.factoryProvider = factoryProvider
        }

        fun build(): ParticleConfig = ParticleConfig(name, sup, factoryProvider)
    }

    fun interface ParticleFactoryProvider {
        fun register(event: RegisterParticleProvidersEvent, particleType: SimpleParticleType)
    }
}
